src/main.js

```diff
@@ -18,85 +20,113 @@ if (!org && org.length === 0) {
-      org: org,
-      package_type: packageType,
-      package_name: packageName,
-      per_page: maxResultsPerPage,
-      page: page
-    });
-
-    if (response.status == 200) {
-      if (response.data) {
-        if (response.data.length < maxResultsPerPage) {
-          hasMorePackages = false;
-        } else {
-          page += 1;
-        }
+const graphqlWithAuth = graphql.defaults({
+  headers: {
+    authorization: `token ${token}`
+  }
+});
 
-        for (let index = 0; index < response.data.length; index++) {
-          const package = response.data[index];
-          if (package.name.indexOf(branchPattern) > -1) {
-            packagesToDelete.push({
-              name: package.name,
-              id: package.id
-            });
-          } else {
-            core.info(`Package ${package.name} does not meet the pattern and will not be deleted`);
+async function getAllPackagesForRepo() {
+  core.info('Querying for all of the packages in the repo.\n');
+  const query = `
+  query {
+    repository(owner: "${org}", name: "${repo}") {
+      packages(packageType: ${packageType.toUpperCase()}, first: 100){
+        totalCount
+        nodes {
+          name
+          id
+          versions(last: 100) {
+            nodes {
+              id,
+              version
+            }
           }
         }
-      } else {
-        core.info('Finished getting packages for the repository.');
-      }
-    } else {
-      core.setFailed(`An error occurred retrieving page ${page} of packages.`);
+      } 
     }
   }
+  `;
+
+  const response = await graphqlWithAuth(query);
+  core.info(`Successfully recieved ${response.repository.packages.totalCount} packages.`);
+  response.repository.packages.nodes.forEach(node => {
+    core.info(node.name);
+  });
+  // Add some space between the list of packages and the following logs.
+  core.info(' ');
+
+  return response.repository.packages.nodes.flatMap(package =>
+    package.versions.nodes.map(version => ({
+      packageName: package.name,
+      versionName: version.version,
+      id: version.id
+    }))
+  );
+}
 
-  if (packagesToDelete.length === 0) {
-    core.info('Finished gathering packages, there were no items to removed.');
-  } else {
-    core.info('Finished gathering packages, the following items will be removed:');
-    console.log(packagesToDelete); //Normally I'd make this core.info but it doesn't print right with JSON.stringify()
+function filterPackagesByName(packages) {
+  if (!packageNames) {
+    return packages;
   }
 
-  return packagesToDelete;
+  const packageNamesArray = packageNames.split(',').map(package => package.trim());
+
+  return packages.filter(package => packageNamesArray.includes(package.packageName));
+}
+
+function filterPackageVersionsByBranchPattern(packages) {
+  const filteredPackages = packages.filter(package => {
+    const shouldDelete = package.versionName.indexOf(branchPattern) > -1;
```

Comment:
This seems like a potentially dangerous filter condition. What if someone accidentally passed a number where they were supposed to pass a branch name? They could end up deleting released packages.

Can we restrict this to searching only in the prerelease label part of the semver version numbers?

Or can GitHub already tell us whether each package is a pre-release or release version, so we can limit the deletion to just the pre-release ones?

Reply:
I should be able to modify the graphql query to tell us if it's a pre-release and limit it to that 👍

Reply:
It turns out we don't set our packages as prerelease. I thought we did, but I was thinking about releases. In order to get packages marked as prerelease, we'd apparently have to suffix our versions with `beta` or something. I'm not sure that's something we want to introduce. I can play around with it if we want. For the time being though, I'm going to work on making the logic look at only the prerelease part of our versions (your first suggestion).

Reply:
So the `PackageVersion` GraphQL object has a `preRelease` boolean property.

I assume it's set automatically by GitHub inspecting the package version as it's published, and that it's based on the presence of a `-` followed by a label in the semver version number of the package being uploaded. Am I wrong?

Reply:
That's how it should work, but GitHub isn't recognizing our version convention as pre-release for some reason. You can see in [this workflow output that `preRelease` is always false](https://github.com/im-practices/cloud-events/runs/6183125661?check_suite_focus=true#step:2:29) (the double false values were because I was printing it out as a string and a boolean so I could verify the value was indeed a boolean). That led me to falsely assume we'd need some special suffix, but after some more digging I don't think that should be the case. I'll continue looking and try to figure out why GitHub isn't treating out pre-release packages as pre-release.

Reply:
Whelp I have no clue why GitHub isn't setting the `preRelease` property like we'd expect. I can't get it to be anything but `false`.

I ended up giving up and adding a [regex check](https://github.com/im-open/delete-branch-package-versions/blob/allow-delete-all-packages-for-repo/src/main.js#L18) to [test a version number for a suffix](https://github.com/im-open/delete-branch-package-versions/blob/allow-delete-all-packages-for-repo/src/main.js#L69).